File: packages/vscode-extension/src/privacy/sanitizer.ts
import { getConfig } from "../config";

export interface RawHeartbeat {
  timestamp: string;
  language?: string;
  project?: string;
  branch?: string;
  isWrite: boolean;
  activeSeconds: number;
  sessionId: string;
  editorName: string;
  os: string;
  status?: "active" | "offline";
}

export function sanitize(hb: RawHeartbeat): RawHeartbeat {
  const cfg = getConfig();
  const sanitized = { ...hb };

  if (!cfg.shareLanguage) {
    delete sanitized.language;
  }

  if (!cfg.shareProject) {
    delete sanitized.project;
    delete sanitized.branch;
  }

  if (sanitized.project && cfg.excludeProjects.length > 0) {
    const lower = sanitized.project.toLowerCase();
    if (cfg.excludeProjects.some((p) => lower.includes(p.toLowerCase()))) {
      delete sanitized.project;
      delete sanitized.branch;
    }
  }

  return sanitized;
}

File: packages/neovim-plugin/lua/gitcity/sanitizer.lua
local M = {}
local config = require("gitcity.config")

function M.sanitize(hb)
  local cfg = config.get()
  local sanitized = vim.deepcopy(hb)
  
  if not cfg.privacy.shareLanguage then
    sanitized.language = nil
  end
  
  if not cfg.privacy.shareProject then
    sanitized.project = nil
    sanitized.branch = nil
  end

  if sanitized.project and cfg.privacy.excludeProjects and #cfg.privacy.excludeProjects > 0 then
    local lower_proj = string.lower(sanitized.project)
    for _, exclude_item in ipairs(cfg.privacy.excludeProjects) do
      if string.find(lower_proj, string.lower(exclude_item), 1, true) then
        sanitized.project = nil
        sanitized.branch = nil
        break
      end
    end
  end

  return sanitized
end

return M
